community/kernel/src/main/java/org/neo4j/kernel/impl/util/ValueUtils.java

```diff
@@ -326,11 +326,25 @@ else if ( map.containsKey( "latitude" ) && map.containsKey( "longitude" ) )
 
     public static NodeValue fromNodeProxy( Node node )
     {
-        return new NodeProxyWrappingNodeValue( node );
+        if ( node instanceof NodeValue )
+        {
+            return (NodeValue) node;
+        }
+        else
+        {
+            return new CoreNodeWrappingNodeValue( node );
+        }
```

Comment:
Even if `NodeProxy` were not to extend `NodeValue` we could still do something special for `NodeProxy` here, and return *sane* `NodeValue` implementations in that case, reserving the unfortunate hack of `NodeValue` with the capability of having behaviour for external implementations of `Node` - which would help us to _limit_ the spread of that disease.